Apple sign-in

lindboe opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

### Fixed Issues
<!---
1. Please replace GH_LINK with a URL link to the GitHub issue this Pull Request is fixing.
2. Please replace PROPOSAL: GH_LINK_ISSUE(COMMENT) with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<number-of-the-issue>
$ https://github.com/Expensify/App/issues/<number-of-the-issue(comment)>

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ https://github.com/Expensify/App/issues/7079  
PROPOSAL: GH_LINK_ISSUE(COMMENT)

### Tests

#### iOS
- [ ] Ensure app builds correctly
- [ ] Select the Apple Icon Button
- [ ] Enter valid Apple ID and Password
- [ ] If prompted, validate with 2 factor authentication
- [ ] Select to share email with app
- [ ] Select continue and sign in with Expensify
- [ ] Ensure logged in correctly

#### Android
- [ ] Ensure app builds correctly
- [ ] Select the Apple Icon Button
- [ ] Enter valid Apple ID and Password
- [ ] If prompted, validate with 2 factor authentication
- [ ] Select to share email with app
- [ ] Select continue and sign in with Expensify
- [ ] Ensure logged in correctly

#### Web
- [ ] Ensure no errors in console
- [ ] Select the Apple Icon Button
- [ ] Ensure Pop Up window appears
- [ ] Enter valid Apple ID and Password
- [ ] If prompted, validate with 2 factor authentication
- [ ] Select to share email with app, if prompted.
- [ ] Select continue and sign in with Expensify
- [ ] Ensure correct redirect to Expensify tab
- [ ] Ensure login occurs

#### Desktop
- [ ] Ensure no errors in console
- [ ] Select the Apple Icon Button
- [ ] Check that is redirects to web tab, url will be `new.expensify.com/sign-in-with-apple` (It will automatically redirect to home login because of redirects)
- [ ] Navigate to `staging.new.expensify.com/sign-in-with-apple`
- [ ] Ensure page loads and there is a `Continue with Apple` button.
- [ ] Select Button
- [ ] Go through web login flow
- [ ] After successful login, ensure there is a prompt to go back to desktop app
- [ ] Select option to navigate back to desktop app
- [ ] Ensure successful navigation
- [ ] Check for successful login once navigated back to desktop app

### Offline tests

Feature is not available in offline mode

### QA Steps

Same as Tests

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Desktop</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android</summary>

<!-- add screenshots or videos here -->

</details>

lindboe commented:
I have read the CLA Document and I hereby sign the CLA

cdanwards commented:
I have read the CLA Document and I hereby sign the CLA

marcochavezf commented on src/libs/actions/Session/index.js:
```diff
@@ -204,55 +204,84 @@ function resendLinkWithValidateCode(login = credentials.login) {
+function handleAppleAuthApiResponse(token) {
+    const {optimisticData, successData, failureData} = generateResponseData();
+    // eslint-disable-next-line rulesdir/no-api-side-effects-method
+    API.makeRequestWithSideEffects('AuthenticateApple', {token}, {optimisticData, successData, failureData})
```
I noticed the backend for this command is expecting `idToken` instead of `token`:

```suggestion
    API.makeRequestWithSideEffects('AuthenticateApple', {idToken: token}, {optimisticData, successData, failureData})
```

silasjmatson replied:
Fixed, ty!

silasjmatson commented:
I have read the CLA Document and I hereby sign the CLA

cdanwards commented:
@marcochavezf I'm unable to edit Lizzi's comment so I'll be posting the videos and testing instructions in comments.

cdanwards commented:
# Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/8878532/b38c4f3e-cca6-4b29-b233-fedd0e745fc6


</details>

<details>
<summary>Mobile Web - Chrome</summary>


https://github.com/Expensify/App/assets/8878532/b2058179-5b1e-4baa-abd7-d74e7edaae15



</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/8878532/80fb588d-695b-404c-8660-74dcb2416bf3


</details>

<details>
<summary>Desktop</summary>


https://github.com/Expensify/App/assets/8878532/64dd9a07-b70c-48b7-bfc7-3783f76c9775



</details>

<details>
<summary>iOS</summary>


https://github.com/Expensify/App/assets/8878532/bbc227f6-53dd-4073-b626-912b37f3c35a



</details>

<details>
<summary>Android</summary>


https://github.com/Expensify/App/assets/8878532/92b2ab38-673f-4523-8d34-cd539889ebc1



</details>

cdanwards commented:
### Tests

#### iOS
- [ ] Ensure app builds correctly
- [ ] Select the Apple Icon Button
- [ ] Enter valid Apple ID and Password
- [ ] If prompted, validate with 2 factor authentication
- [ ] Select to share email with app
- [ ] Select continue and sign in with Expensify
- [ ] Ensure logged in correctly

#### Android
- [ ] Ensure app builds correctly
- [ ] Select the Apple Icon Button
- [ ] Enter valid Apple ID and Password
- [ ] If prompted, validate with 2 factor authentication
- [ ] Select to share email with app
- [ ] Select continue and sign in with Expensify
- [ ] Ensure logged in correctly

#### Web
- [ ] Ensure no errors in console
- [ ] Select the Apple Icon Button
- [ ] Ensure Pop Up window appears
- [ ] Enter valid Apple ID and Password
- [ ] If prompted, validate with 2 factor authentication
- [ ] Select to share email with app, if prompted.
- [ ] Select continue and sign in with Expensify
- [ ] Ensure correct redirect to Expensify tab
- [ ] Ensure login occurs

#### Desktop
- [ ] Ensure no errors in console
- [ ] Select the Apple Icon Button
- [ ] Check that is redirects to web tab, url will be `new.expensify.com/sign-in-with-apple` (It will automatically redirect to home login because of redirects)
- [ ] Navigate to `staging.new.expensify.com/sign-in-with-apple`
- [ ] Ensure page loads and there is a `Continue with Apple` button.
- [ ] Select Button
- [ ] Go through web login flow
- [ ] After successful login, ensure there is a prompt to go back to desktop app
- [ ] Select option to navigate back to desktop app
- [ ] Ensure successful navigation
- [ ] Check for successful login once navigated back to desktop app

### Offline tests

Feature is not available in offline mode

### QA Steps

Same as Tests

marcochavezf commented:
> @marcochavezf I'm unable to edit Lizzi's comment so I'll be posting the videos and testing instructions in comments.

Ok, I will update the OP with the instructions.

marcochavezf commented:
Since we're rolling this out alongside a new feature allowing [unauthenticated users to open public rooms](https://github.com/Expensify/App/pull/18293), could you also run through these steps while testing? @cdanwards @rushatgabhane 

1. Sign out
2. Open a public room (like the [ExpensiConX #announce room](https://staging.new.expensify.com/r/3504895439653267) or this [other public room](https://staging.new.expensify.com/r/6469015542180863) for testing)
3. Verify you can see the announce room as an anonymous user (random email).
4. Click on `Sign in` or any action that should not be allowed to an unauthenticated user
5. Verify you're redirected to the signin page
6. Enter email/phone
7. Enter magic code
8. Verify you can see the public room in the LHN
9. Repeat steps 1-5 
10. Sign-in with Apple
11. Verify you can still see the public room in the LHN

If you find any issue let me know so I can fix it immediately

cdanwards commented:
@rushatgabhane just pushed those updates! Let me know if there's anything else!

marcochavezf commented:
Hi @0xmiroslav, any update on the review?

lindboe commented on src/Expensify.js:
```diff
@@ -181,7 +180,7 @@ function Expensify(props) {
 
     return (
-        <DeeplinkWrapper>
+        <>
```
> This will cause regression of not showing open desktop app popup on unauthorized screens (i.e. login page, public chat room)

This is intended. We asked in Slack and there was no indication that the prompt to open the app in desktop prior to sign-in was essential: https://expensify.slack.com/archives/C01GTK53T8Q/p1683070404415199

Sorry that wasn't conveyed more clearly.

lindboe replied:
Follow up from Slack thread: It looks like this should still work, because `authToken` is still set even in this "anonymous user" use case (https://expensify.slack.com/archives/C01GTK53T8Q/p1686181191875739?thread_ts=1683070404.415199&cid=C01GTK53T8Q)

cdanwards commented on src/components/SignInButtons/GetUserLanguage.js:
```diff
@@ -0,0 +1,14 @@
+const localeCodes = {
```
Following up on this comment. After reviewing the `preferredLocale` prop provided by `withLocalize`, it seems the format of that prop isn't sufficient for this use case. That prop only returns `es` or `en`, where we need the country locale formatting for the apple script, i.e. : `en_US` or `es_US`. Maybe it might be better to give this component a different title so that the intent is communicated better?

lindboe commented on src/pages/signin/SignInPageLayout/index.js:
```diff
@@ -59,6 +61,10 @@ const SignInPageLayout = (props) => {
 
     useEffect(scrollPageToTop, [props.welcomeHeader, props.welcomeText]);
 
+    if (!isFocused) {
+        return null;
+    }
+
```
For my own education, why do we use an HOC? I thought we were changing to hooks with functional components?

cdanwards commented:
@marcochavezf @0xmiroslav  We've updated this PR to address the comments from the reviews and have updated it with the latest changes from `main`. I believe it's ready to be merged unless there are any other concerns!

marcochavezf commented:
Hi @shawnborton, can we get your design review based on the videos [here](https://github.com/Expensify/App/pull/16723#issuecomment-1570405485)?

shawnborton commented:
Flying with bad wifi today, but will try to take a look later tonight or tomorrow morning.

shawnborton commented:
This looks good to me, nice job!

lindboe commented:
@0xmiroslav the default dev configuration for web for Apple sign-in does not work because Apple does not allow `localhost` as the origin of the sign-in flow. It requires we use an `https` domain for the web app, that's been registered with the service ID we're using in the configuration of the web library.

That is what the `contributingGuides` instructions are for: https://github.com/Expensify/App/pull/16723/files#diff-f48fd1feb66bd44dc65951c62e27d2b0144a6ed50436212f694cfd0658cfe5ccR7.

They show how to accomplish testing the flow in two parts:

1. Hosting the dev app on an https domain, using ssh tunneling, where that domain has been configured to work with an Apple service ID, will allow you to finish the flow to the point of closing the pop-up window and receiving a token from Apple. This requires setting up your own Apple service ID, instead of Expensify's, so that you can configure it to allow your specific domain that you're serving the dev app on.
2. However, since this Apple token wasn't configured with Expensify's service ID, trying to log in with it won't work. To get around this, we've been generating and intercepting Apple tokens on Android, then using an environment variable to override the value used on web: https://github.com/Expensify/App/pull/16723/files#diff-f48fd1feb66bd44dc65951c62e27d2b0144a6ed50436212f694cfd0658cfe5ccR33

It's unfortunate it's this complicated in the development environment, but this all should work normally once we're in staging or prod, as those are hosted on https domains which the Expensify service ID has been configured to use.

We could've worked around the origin domain requirement by using a redirect-style flow instead of a pop-up, but that had additional complications that were hard for us to design and test without access to the backend. Perhaps the internal Expensify team could look into this in the future.

lindboe commented:
> Is web testable on PR build generated after adding Ready To Build label? It's https.

Does this build for the staging environment on web? Sorry, I'm not familiar with this process yet. (cc @marcochavezf )

marcochavezf commented:
> Is web testable on PR build generated after adding Ready To Build label?

Ah good question, I don't know. I triggered the actions for the ad-hoc builds, we'll see

OSBotify commented:
Use the links below to test this build in android and iOS. Happy testing!

OSBotify commented:
:test_tube::test_tube: Use the links below to test this build in android and iOS. Happy testing! :test_tube::test_tube:
| android :robot:  | iOS :apple: |
| ------------- | ------------- |
| https://ad-hoc-expensify-cash.s3.amazonaws.com/android/16723/index.html  | https://ad-hoc-expensify-cash.s3.amazonaws.com/ios/16723/index.html  |
| ![Android](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://ad-hoc-expensify-cash.s3.amazonaws.com/android/16723/index.html)  | ![iOS](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://ad-hoc-expensify-cash.s3.amazonaws.com/ios/16723/index.html)  |
| desktop :computer: | web :spider_web: |
| https://ad-hoc-expensify-cash.s3.amazonaws.com/desktop/16723/NewExpensify.dmg  | https://16723.pr-testing.expensify.com  |
| ![Desktop](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://ad-hoc-expensify-cash.s3.amazonaws.com/desktop/16723/NewExpensify.dmg)  | ![Web](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://16723.pr-testing.expensify.com)  |

lindboe commented:
Okay yeah, so since these generate dynamic domain names (`https://16723.pr-testing.expensify.com/`), testing on them won't work either, unless we had permission to go update the Expensify Apple config when they're generated (which we don't).

If we could host them on a reserved static domain name for this purpose (e.g., `https://apple.pr-testing.expensify.com`), we could put that in the config once and not think about it anymore, but obviously that doesn't scale well if multiple commits need to make use of it.

lindboe commented on package-lock.json:
```diff
@@ -17,7 +17,8 @@
         "@gorhom/portal": "^1.0.14",
-        "@oguzhnatly/react-native-image-manipulator": "github:Expensify/react-native-image-manipulator#5cdae3d4455b03a04c57f50be3863e2fe6c92c52",
+        "@invertase/react-native-apple-authentication": "^2.2.2",
+        "@oguzhnatly/react-native-image-manipulator": "github:Expensify/react-native-image-manipulator#c5f654fc9d0ad7cc5b89d50b34ecf8b0e3f4d050",
```
I'm looking at this and am concerned about the number of merges/conflict resolutions on this branch. I'm considering whether re-constructing it off of main and re-signing commits (and a force-push) might be the safest approach. @0xmiroslav would that cause any problems for you?

cdanwards replied:
This has been reverted.

lindboe commented:
@AndrewGable has edited the Apple sign-in config for us before. Can we add the PR domain (`https://16723.pr-testing.expensify.com/`) and a corresponding redirect URL (`https://16723.pr-testing.expensify.com/appleauth`, does not have to exist, isn't actually used) to the `com.chat.expensify.chat.AppleSignIn` service ID config temporarily? (cc @marcochavezf )

marcochavezf commented:
> @AndrewGable has edited the Apple sign-in config for us before. Can we add the PR domain (`https://16723.pr-testing.expensify.com/`) and a corresponding redirect URL (`https://16723.pr-testing.expensify.com/appleauth`, does not have to exist, isn't actually used) to the `com.chat.expensify.chat.AppleSignIn` service ID config temporarily? (cc @marcochavezf )

Yeah, given the complexity of configuring Apple sign-in for local testing for web, it will be fine to test it once it's on staging. If it's a recurrent problem to test/troubleshoot issues locally, we could look for other ways to make it easier cc @0xmiroslav 

@lindboe @cdanwards could you resolve and/or comment on the pending review comments?

lindboe commented:
@marcochavezf Yup, Dan is testing out the changes and we'll have them up to review shortly!

cdanwards commented on src/pages/signin/SignInPage.js:
```diff
@@ -150,7 +150,7 @@ class SignInPage extends Component {
             // There is an issue SafeAreaView on Android where wrong insets flicker on app start.
             // Can be removed once https://github.com/th3rdwave/react-native-safe-area-context/issues/364 is resolved.
-            <View style={[styles.signInPage, StyleUtils.getSafeAreaPadding(this.props.insets, 1)]}>
+            <View style={[styles.signInPage, StyleUtils.getSafeAreaPadding({...this.props.insets, bottom: 0}, 1)]}>
```
I'm adding in a comment for clarity, but the design asked that the `SignInPage` svg (snakes and hands) to pass under the Home Indicator. This wasn't an issue before the Apple sign in button, but on smaller screen the apple sign in button pushed the svg so that it was cut off by the safe area inset instead of flowing under the Home Indicator.

cdanwards commented on src/pages/signin/SignInPageLayout/SignInPageContent.js:
```diff
@@ -82,6 +82,11 @@ function SignInPageContent(props) {
                         </View>
                     ) : null}
                 </View>
+                {props.isSmallScreenWidth ? (
+                    <View>
+                        <SignInHeroImage />
+                    </View>
+                ) : null}
```
Removing this.

cdanwards commented on src/pages/signin/SignInPageLayout/index.js:
```diff
@@ -59,6 +60,10 @@ function SignInPageLayout(props) {
 
     useEffect(scrollPageToTop, [props.welcomeHeader, props.welcomeText]);
 
+    if (!props.isFocused) {
+        return null;
+    }
+
```
After testing, this seems to be redundant. Removing this.

cdanwards commented on src/pages/signin/ThirdPartySignInPage.js:
```diff
@@ -0,0 +1,75 @@
+    ...windowDimensionsPropTypes,
+};
+
+const capitalize = (word) => word.charAt(0).toUpperCase() + word.slice(1);
```
Removing this.

cdanwards commented on src/components/SignInButtons/ButtonBase.js:
```diff
@@ -0,0 +1,34 @@
+import React from 'react';
```
This has been updated to `IconButton`

cdanwards commented on src/components/SignInButtons/AppleSignIn/index.android.js:
```diff
@@ -0,0 +1,60 @@
+import * as Session from '../../../libs/actions/Session';
+import CONST from '../../../CONST';
+
+const appleLogoIcon = require('../../../../assets/images/signIn/apple-logo.svg').default;
```
Originally this was to prevent a failing test. Instead, the logo was moved to Expensicons because those are mocked correctly for jest.

cdanwards commented:
@marcochavezf @0xmiroslav the comments from the latest review have either been answered or resolved in the most recent commits. Let me know if there are any questions!

cdanwards commented:
@0xmiroslav are there any questions you have that we can answer to help move this PR along? Thanks!

cdanwards commented:
@0xmiroslav 👋🏼  Hey! Do you have an update on when your review might be in? We're planning to iterate on these changes, but if it might be a while we're considering including the iteration in this PR. Thanks!

marcochavezf commented:
Hi @0xmiroslav, is there an update on the review? Since the Performance Tests GH action is caused by the package-lock.json, I think we can continue with the review

OSBotify commented:
:test_tube::test_tube: Use the links below to test this build in android and iOS. Happy testing! :test_tube::test_tube:
| android :robot:  | iOS :apple: |
| ------------- | ------------- |
| https://ad-hoc-expensify-cash.s3.amazonaws.com/android/16723/index.html  | https://ad-hoc-expensify-cash.s3.amazonaws.com/ios/16723/index.html  |
| ![Android](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://ad-hoc-expensify-cash.s3.amazonaws.com/android/16723/index.html)  | ![iOS](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://ad-hoc-expensify-cash.s3.amazonaws.com/ios/16723/index.html)  |
| desktop :computer: | web :spider_web: |
| https://ad-hoc-expensify-cash.s3.amazonaws.com/desktop/16723/NewExpensify.dmg  | https://16723.pr-testing.expensify.com  |
| ![Desktop](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://ad-hoc-expensify-cash.s3.amazonaws.com/desktop/16723/NewExpensify.dmg)  | ![Web](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://16723.pr-testing.expensify.com)  |

cdanwards commented:
Hey @0xmiroslav @marcochavezf is there anything we can do to help get this review rolling again? Thanks!

cdanwards commented:
@0xmiroslav 👍🏼 Sounds like a plan. I just tested locally on web/desktop/mWeb just to double check nothing from `main` broke anything and it is all working as expected.

shawnborton commented:
Hmm we shouldn't be touching the width there - any idea what happened?

cdanwards commented:
@0xmiroslav @shawnborton taking a look at this.

lindboe commented on src/styles/styles.js:
```diff
@@ -1155,20 +1155,20 @@ const styles = {
     },
 
     signInPageLeftContainer: {
-        paddingLeft: 40,
```
Hmm, okay, I think this came from implementing the desktop-specific pages, re-using components, but only referencing the desktop-specific design Figma: https://www.figma.com/file/pLyF0knXQdbHflO4ILhssB/Google-and-Apple-Sign-In?type=design&node-id=1-13261&mode=design&t=f2oZnbeU9Qucu3Ai-0

In that one, the button (and it's container) is 279px wide. In the other Figma I have a link to, the button/container are 303px wide: https://www.figma.com/file/zPqFGweGMhtmbznXQJdWAU/New-Home-Page?type=design&node-id=144-24152&mode=design&t=iPzNL9YoO5rlDdrx-0.

@shawnborton should we just stick to the width that exists on production?

@cdanwards if we change the width we'll also need to change the width in the web button. We should probably pull that out into a variable in the styles file.

shawnborton replied:
Yeah, let's just stick with whatever we have in production.

marcochavezf commented on contributingGuides/TESTING_APPLE_GOOGLE_SIGNIN.md:
```diff
@@ -0,0 +1,128 @@
+
+### Web
+
+The Sign in with Apple process will break after the user signs in if the pop-up process is not started from a page at an HTTPS domain registered with Apple. We can make a new configuration with your custom HTTPS domain, but then the Apple configuration won't match that of Expensify's backend.
```
Minor change here, since contributors will view this documentation, it could create confusion if it implies that Expensify will make the configuration for them.

lindboe replied:
Good shout, addressed this

lindboe commented:
Made the requested changes, and while fixing the desktop layout, I realized the "Sign in with" text was kind of small on web; I re-checked the Figma and I think we actually wanted font size 13, where this other style was 11, so I swapped the text style it was using.

Updated layout vs. prod:

![image](https://github.com/Expensify/App/assets/5252268/f00a5ea4-9f0a-48fd-9b55-a41d8f19102d)

lindboe commented:
Pushed accessibilityLabel fixes to address the lint issue. Haven't tested screen readers yet, but I'm pretty confident in this change. I'll run those tests on device builds tomorrow.

shawnborton commented:
Sounds good re: updating to 13, I agree that is expected.

OSBotify commented:
:test_tube::test_tube: Use the links below to test this build in android and iOS. Happy testing! :test_tube::test_tube:
| android :robot:  | iOS :apple: |
| ------------- | ------------- |
| https://ad-hoc-expensify-cash.s3.amazonaws.com/android/16723/index.html  | https://ad-hoc-expensify-cash.s3.amazonaws.com/ios/16723/index.html  |
| ![Android](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://ad-hoc-expensify-cash.s3.amazonaws.com/android/16723/index.html)  | ![iOS](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://ad-hoc-expensify-cash.s3.amazonaws.com/ios/16723/index.html)  |
| desktop :computer: | web :spider_web: |
| https://ad-hoc-expensify-cash.s3.amazonaws.com/desktop/16723/NewExpensify.dmg  | https://16723.pr-testing.expensify.com  |
| ![Desktop](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://ad-hoc-expensify-cash.s3.amazonaws.com/desktop/16723/NewExpensify.dmg)  | ![Web](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://16723.pr-testing.expensify.com)  |

marcochavezf commented:
> Please follow step in this video. Getting many console errors in IconButton component

Oh thanks for catching that error, since there's going to be another PR for desktop we can fix it there

0xmiroslav reviewed (approved):
iOS / android looks good.
Web based platforms (web / desktop / mChrome / mSafari) tests will be done in staging build

marcochavezf reviewed (dismissed):
Perfect, thanks @0xmiroslav! It also LGTM, I will merge it and test the remaining platforms in staging

marcochavezf commented:
@cdanwards @lindboe oops, we can't merge because some commits are not signed/verified, could you fix that?

<img width="375" alt="Screenshot 2023-07-13 at 17 13 25" src="https://github.com/Expensify/App/assets/6829422/c77a0f1f-473d-4c95-bf84-90ab8e92f527">

cdanwards commented:
@0xmiroslav fixed the conflicts. @marcochavezf taking a look at the unsigned commits now

lindboe committed:
Apple sign in

Adds Apple sign-in (and a bit of early support for Google) for all
platforms.

Commit squashed to fix signing issues.

cdanwards commented:
@marcochavezf looks like there might be a cache issue on the actions that are failing? Would you be able to rerun those checks?

marcochavezf commented:
> Hey, so to address the signing issue, we decided to squash the commits into a new one. Going back and re-signing involved going through a lot of merge conflicts again and we weren't sure it was worth the time; but if this is a problem we can always go back and go through that again.

Ah we would like to keep the GH history for accountability and transparency, sorry for asking this given the extra effort required to fix the conflicts, but is there another way to keep the GH history?

lindboe commented:
There's no way to fix the bad signature without altering history in some way, and re-signing the commits this PR adds as someone still on the team. The options are:

1. Squash the problem commit into later commits, and re-author and re-sign (like we did here)
2. Rebase: One of us on the team will re-author and re-sign the problem commits and all commits after it (as changing the earlier commits changes the tree and the later signatures become invalid). This requires us to go back and re-fix all of the existing merge conflicts (we do have git rerere set up, but that isn't catching everything)
3. Re-construct: Rebuild relevant changes on top of a more recent copy of main. This will also not keep the original commits exactly, but we could mostly maintain the changesets and titles of each, if that helps (although any that were built against older versions of `main` and later fixed in a merge conflict resolution would be changed).

Can you tell us more about the policy of keeping GH history for accountability and transparency? It would help us figure out a path forward for us to understand what the goals are, since we can't keep the GH history exactly the same while also fixing this issue.

marcochavezf commented:
Hi @lindboe, thanks for bringing up different options, we use the number of commits as an internal metric for external contributions. If there's a way to keep roughly the original number of commits that would be fine (it's not necessary to keep the original author).

lindboe commented:
We can do that. Just to make sure it's clear, I'm a little concerned that it will both 1. Take more time, and 2. Carry more risk of accidentally introducing a change (hence adding to the time it takes to test that everything).

Options #2 and #3 require the branch updater to make a lot of choices about what to pull in and how to fix conflicts, and it's easier to make mistakes.

By contrast, the squashed changeset for #1 was generated from the diff between the last merged main commit and the HEAD of the PR branch, so it's more guaranteed to be the same as what was reviewed.

Just checking to make sure this is the tradeoff we want. We'll get it started.

cdanwards committed:
Merge branch 'main' into apple-sign-in

marcochavezf commented:
Given the complexity of fixing the conflicts and all the extra work that implies restoring the history of the commits, it makes sense to go on with the squashed commit. Thank you for providing context and explaining the tradeoffs 🚀

marcochavezf commented:
There's a linting error:

```
Run npm run prettier -- --loglevel silent

> new.expensify@1.3.42-[10](https://github.com/Expensify/App/actions/runs/5588001071/jobs/10214707775?pr=16723#step:6:11) prettier
> prettier --write . "--loglevel" "silent"

src/styles/styles.js
Error: Prettier diff detected! Please run `npm run prettier` and commit the changes.
Error: Process completed with exit code 1.
```

cdanwards committed:
Fixed linting error

marcochavezf commented:
Oops conflicts again

cdanwards committed:
Fixed merge conflicts from main

cdanwards commented:
@marcochavezf I fixed those conflicts!

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5602380038) to staging by https://github.com/marcochavezf in version: 1.3.43-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌ 
🕸 web 🕸|success ✅

ayazhussain79 commented:
Apps not working getting this error

![Screenshot 2023-07-19 at 10 36 55 PM (1)](https://github.com/Expensify/App/assets/56740856/112bc975-7181-4a6a-9cf7-552e85651bbb)

ayazhussain79 commented:
yes i have

marcochavezf commented:
Hi @cdanwards @lindboe the PR was deployed to staging and seems there's an issue with CSP to load the apple script

<img width="1083" alt="Screenshot 2023-07-19 at 16 32 44" src="https://github.com/Expensify/App/assets/6829422/22444e88-ce02-4893-8c40-2bb95ed883ca">

lindboe commented:
@marcochavezf as far as I'm aware, that's a backend change to add the script host (`https://appleid.cdn-apple.com`) to the `script-src` directive, right?

mvtglobally commented:
@lindboe @marcochavezf @rushatgabhane @shawnborton We are facing the issue with the login. Should we log a bug, or its expected now? 

https://github.com/Expensify/App/assets/43995119/87d906cf-9e4a-4d9e-b8c3-a625044e4b7a

marcochavezf commented:
Hmm this a bug but I think we should need to revert until we fix the CSP issue. @roryabraham given how the app is hosted for web, I think we'd to wait until NewDot is migrated to our servers to fix the CSP issue, correct? Or do you know if there's way to allow load the apple script for this case?

roryabraham commented:
Depending on the urgency we can fix the existing CSP in CloudFlare.

roryabraham commented:
I think we can make this change in the CSP pretty easily, I'll create a PR...

roryabraham commented:
@marcochavezf https://github.com/Expensify/Cloudflare-Workers/pull/67

FWIW, static things are pretty easy to add to our CSP, but dynamic things (like injecting a nonce into script tags) are harder. The main thing is that we don't currently have any way to locally test our CSP, and we're working on fixing that by moving NewDot web hosting from S3 to our servers and re-implementing the CSP in PHP instead of in a CloudFlare worker where it lives today.

marcochavezf commented:
Perfect, thank you! We'd just need to wait for the PR that updates the CSP in the Cloudflare worker

robertjchen commented:
CSP was updated 👍

marcochavezf commented:
@cdanwards @lindboe seems the authorization request to the Apple servers is successful, but it seems it's not redirecting/sending the response to the site:

<img width="1219" alt="Screenshot 2023-07-20 at 11 35 36" src="https://github.com/Expensify/App/assets/6829422/03f2621c-4430-488c-ba70-c3e7424d424f">

Looking at the URL seems we're redirecting to the production site (new.expensiy.com vs staging (staging.new.expensify.com): 

```
https://appleid.apple.com/auth/authorize?client_id=com.chat.expensify.chat.AppleSignIn&redirect_uri=https%3A%2F%2Fnew.expensify.com%2Fappleauth&response_type=code%20id_token&scope=name%20email&response_mode=web_message&frame_id=782dc751-a25b-48fe-a85e-00401bd3cb9e&m=12&v=1.5.5
```

marcochavezf commented:
Linking here an issue where we were using `DeeplinkWrapper` in `PublicScreens` https://github.com/Expensify/App/issues/23285#issuecomment-1644197746

marcochavezf commented:
> Linking here an issue where we were using DeeplinkWrapper in PublicScreens https://github.com/Expensify/App/issues/23285#issuecomment-1644197746

@cdanwards @lindboe could you check this issue ^? I wonder if we can move `DeeplinkWrapper` to where it was before

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5602380038) to staging by https://github.com/marcochavezf in version: 1.3.43-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌ 
🕸 web 🕸|success ✅

iwiznia commented:
This caused a regression, so reverting it https://github.com/Expensify/App/issues/23285

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5625378789) to staging by https://github.com/marcochavezf in version: 1.3.44-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5625371664) to production by https://github.com/jasperhuangg in version: 1.3.43-7 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5651460665) to production by https://github.com/marcaaron in version: 1.3.44-2 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
